File: leetcode/201-300/218-the-skyline-problem/skyline.py
"""LeetCode 218 — The Skyline Problem (Python mirror / oracle).

Divide and conquer (merge-sort shaped): the skyline of one building is
[[L,H],[R,0]]; merge two contours by sweeping both and emitting a key point
whenever the max of the two current heights changes. Mirrors the Kara version.
"""


def merge(left, right):
    result = []
    nl, nr = len(left), len(right)
    i = j = 0
    h1 = h2 = 0
    while i < nl and j < nr:
        if left[i][0] < right[j][0]:
            x = left[i][0]
            h1 = left[i][1]
            i += 1
        elif left[i][0] > right[j][0]:
            x = right[j][0]
            h2 = right[j][1]
            j += 1
        else:
            x = left[i][0]
            h1 = left[i][1]
            h2 = right[j][1]
            i += 1
            j += 1
        maxh = h1 if h1 > h2 else h2
        if not result or result[-1][1] != maxh:
            result.append([x, maxh])
    while i < nl:
        result.append(left[i])
        i += 1
    while j < nr:
        result.append(right[j])
        j += 1
    return result


def skyline(bs, lo, hi):
    if hi - lo == 1:
        l, r, h = bs[lo]
        return [[l, h], [r, 0]]
    mid = lo + (hi - lo) // 2
    left = skyline(bs, lo, mid)
    right = skyline(bs, mid, hi)
    return merge(left, right)


def report(bs):
    if not bs:
        print("[]")
        return
    sky = skyline(bs, 0, len(bs))
    print(" ".join(f"[{p[0]},{p[1]}]" for p in sky))


def main():
    report([[2, 9, 10], [3, 7, 15], [5, 12, 12], [15, 20, 10], [19, 24, 8]])
    report([[0, 2, 3], [2, 5, 3]])
    report([[1, 5, 11]])
    report([[1, 2, 1], [1, 2, 2], [1, 2, 3]])
    report([[1, 3, 4], [5, 7, 6]])
    report([[0, 10, 5], [3, 6, 3]])
    report([])


main()

File: leetcode/201-300/218-the-skyline-problem/bench/skyline.rs
#[derive(Clone, Copy)]
struct Building {
    left: i64,
    right: i64,
    height: i64,
}

#[derive(Clone, Copy)]
struct Point {
    x: i64,
    h: i64,
}

fn merge(left: &[Point], right: &[Point]) -> Vec<Point> {
    let mut result: Vec<Point> = Vec::new();
    let nl = left.len() as i64;
    let nr = right.len() as i64;
    let mut i = 0i64;
    let mut j = 0i64;
    let mut h1 = 0i64;
    let mut h2 = 0i64;
    while i < nl && j < nr {
        let x;
        if left[i as usize].x < right[j as usize].x {
            x = left[i as usize].x;
            h1 = left[i as usize].h;
            i += 1;
        } else if left[i as usize].x > right[j as usize].x {
            x = right[j as usize].x;
            h2 = right[j as usize].h;
            j += 1;
        } else {
            x = left[i as usize].x;
            h1 = left[i as usize].h;
            h2 = right[j as usize].h;
            i += 1;
            j += 1;
        }
        let maxh = if h1 > h2 { h1 } else { h2 };
        if result.is_empty() || result[result.len() - 1].h != maxh {
            result.push(Point { x, h: maxh });
        }
    }
    while i < nl {
        result.push(left[i as usize]);
        i += 1;
    }
    while j < nr {
        result.push(right[j as usize]);
        j += 1;
    }
    result
}

fn skyline(bs: &[Building], lo: i64, hi: i64) -> Vec<Point> {
    if hi - lo == 1 {
        let b = bs[lo as usize];
        return vec![Point { x: b.left, h: b.height }, Point { x: b.right, h: 0 }];
    }
    let mid = lo + (hi - lo) / 2;
    let left = skyline(bs, lo, mid);
    let right = skyline(bs, mid, hi);
    merge(&left, &right)
}

fn main() {
    let n: i64 = 24000;
    let passes: i64 = 100;
    let xr: i64 = 200000;
    let wr: i64 = 120;
    let hr: i64 = 1000;

    let mut bs: Vec<Building> = Vec::new();
    let mut state: i64 = 12345;
    for _ in 0..n {
        state = (state * 1103515245 + 12345) & 2147483647;
        let left = state % xr;
        state = (state * 1103515245 + 12345) & 2147483647;
        let width = 1 + state % wr;
        state = (state * 1103515245 + 12345) & 2147483647;
        let height = 1 + state % hr;
        bs.push(Building { left, right: left + width, height });
    }

    let mut sink: i64 = 0;
    for p in 0..passes {
        state = (state * 1103515245 + 12345) & 2147483647;
        let idx = (p % n) as usize;
        bs[idx].height = 1 + state % hr;
        let sky = skyline(&bs, 0, n);
        for pt in &sky {
            sink += pt.x + pt.h;
        }
    }
    println!("{}", sink);
}
